### ISSUE TITLE
out of sample forecast

### ISSUE THREAD

**Issue by waudinio27** [sydney-machine-learning/Bayesianneuralnetworks-MCMC-tutorial#17]
I really like the notebook number 4. I wanted to ask if it is possible to show a out of sample forecast for several steps after the test data? 
Like with an additional function or something. 

I struggle to implement it myself. 

Another thing would be various chains in parallel to speed up sampling - but this is maybe out of scope. 

Thanks a lot

**Comment by rohitash-chandra** [comment]
code for parallel chains in repo of this paper: Chandra, R., Jain, K., Deo, R. V., & Cripps, S. (2019). Langevin-gradient parallel tempering for Bayesian neural learning. Neurocomputing, 359, 315-326.

**Comment by waudinio27** [comment]
Thanks a lot Dr. Chandra :-) 

I will take myself some time to dive into the paper and trying to adopt it! 

What would you say about my forecast function using the forward.pass - that I am trying to  create? Is it somehow on the right track? 

import numpy as np

def forecast_next_days(model, last_data, days=15, scaler=None):
    """
    Forecast future values using a trained neural network model without a fixed sequence length.
    
    Parameters:
    - model: The trained NeuralNetwork model.
    - last_data: The most recent data point to start forecasting.
    - days: Number of days to forecast into the future.
    - scaler: The scaler used to normalize the data (if any).

    Returns:
    - Forecasts: Array of forecasted values for the next 'days'.
    """
    predictions = []
    data = np.copy(last_data)  # Start with the last available data point

    for _ in range(days):
        # Reshape data for the model
        data_input = data.reshape(1, -1)
        
        # Predict the next value
        prediction = model.forward_pass(data_input)
        predicted_value = prediction[0, -1]  # Assuming the last output node is the prediction
        
        predictions.append(predicted_value)
        
        # Update data for the next prediction
        data = np.roll(data, shift=-1)
        data[-1] = predicted_value

        print(f"Step {_+1}, Predicted Value: {predicted_value}, Updated Data: {data}")

    # If a scaler was used for normalization, inverse transform the predictions
    if scaler:
        predictions = scaler.inverse_transform(np.array(predictions).reshape(-1, 1))

    return np.array(predictions)

# Example usage:
num_future_days = 15
forecast = forecast_next_days(nn_model, x_test[-1], days=num_future_days, scaler=None)

print("Forecast for the next 15 days:", forecast)

Best regards

Matthias
